### PR TITLE
Refactor batch sample matching

### PR DESCRIPTION
**Pull request by benjeffery**
Refactor to work like ancestor matching.
Some feedback on the approach would be good here.

### CODE REVIEW

File: tsinfer/inference.py [hunk 1]
```diff
@@ -904,6 +907,173 @@ def augment_ancestors(
...
+    metadata["sample_times"] = sample_times
+    num_samples_per_partition = min_work_per_job // variant_data.num_sites
+    if num_samples_per_partition == 0:
+        num_samples_per_partition = 1
+    metadata["num_samples_per_partition"] = num_samples_per_partition
+    metadata["num_partitions"] = math.ceil(
+        len(sample_indexes) / num_samples_per_partition
+    )
+    metadata_path = work_dir / "metadata.json"
+    metadata_path.write_text(json.dumps(metadata))
```
Comment (jeromekelleher) [inline tsinfer/inference.py:1014]: handy to indent this too for readability.

File: tsinfer/inference.py [hunk 2]
```diff
@@ -904,6 +907,173 @@ def augment_ancestors(
...
+    metadata["sample_indexes"] = sample_indexes
+    metadata["sample_times"] = sample_times
+    num_samples_per_partition = min_work_per_job // variant_data.num_sites
+    if num_samples_per_partition == 0:
+        num_samples_per_partition = 1
+    metadata["num_samples_per_partition"] = num_samples_per_partition
+    metadata["num_partitions"] = math.ceil(
+        len(sample_indexes) / num_samples_per_partition
+    )
+    metadata_path = work_dir / "metadata.json"
```
Comment (jeromekelleher) [inline tsinfer/inference.py:1013]: We can keep the "metadata.json" name in the file path, that's fine I think.

File: tsinfer/inference.py [hunk 3]
```diff
@@ -904,6 +907,173 @@ def augment_ancestors(
...
+    if max_num_partitions is None:
+        max_num_partitions = 1000
+
+    # Convert working_dir to pathlib.Path
+    work_dir = pathlib.Path(work_dir)
+
+    # Create work dir
+    work_dir.mkdir(parents=True, exist_ok=True)
+
+    metadata = {
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply tsinfer/inference.py]: Fixed in https://github.com/tskit-dev/tsinfer/pull/954/commits/69249298f85f8e80156151dc48c8f12d36387e40

File: tsinfer/inference.py (hunk 1 above)
Comment (benjeffery) [reply tsinfer/inference.py]: Fixed in https://github.com/tskit-dev/tsinfer/pull/954/commits/6a833bbd5b2f3cd61eea7fd0d43864f3aaf32a99

**Review by jeromekelleher — APPROVED** [review]
Looks good, happy to merge

**Pull request merged**